### ISSUE TITLE
Consider removing `R` as it is always the identity matrix

### ISSUE THREAD

**Issue by dafeda** [equinor/iterative_ensemble_smoother#64]
`R` is not needed when doing exact inversion which is what we use most of the time.
It's by default defined as `R = np.identity(len(observation_errors), dtype=np.double)` so it can take up a bit of memory if there are lots of observations.

If at some point we manage to model the covariance between observations, we would have to re-introduce it.

**Comment by geirev** [comment]
It makes sense to remove R as a matrix, and then also the inversion options that use it.
If we later introduce measurement error correlations, it is better (faster and less memory-consuming) to represent them through the measurement error perturbations in the matrix E (m x N) and use that R ~ E E^T.

**Comment by dafeda** [comment]
> It makes sense to remove R as a matrix, and then also the inversion options that use it. If we later introduce measurement error correlations, it is better (faster and less memory-consuming) to represent them through the measurement error perturbations in the matrix E (m x N) and use that R ~ E E^T.

That makes sense, thank you.

**Comment by dafeda** [comment]
I have created a PR (https://github.com/equinor/iterative_ensemble_smoother/pull/67) that removes the possibility of using exact `R` as well as `R` estimated using low-rank `E`.
It has not really been possible to pass an exact `R` anyway and I doubt anyone has used the low rank `E` option.

We now only have exact inversion in which case we do not create `R` at all, and subspace inversion using `EE^T`.

**Comment by geirev** [comment]
An afterthought is that as the iterative smoother is now a separate repository, I would suggest keeping the different functionalities (except the one with R estimated from a low-rank E), as other users might want them. Maybe it is better to disable them from the ERT menu to make them invisible to the users and not allocate space for R in ERT? After all, it is not a lot of extra code.

We will use the subspace inversion with a low-rank E next year as we are starting a project using it with ERT.

**Comment by dafeda** [comment]
You write "except the one with R estimated from a low-rank E" but also that we will use subspace inversion with a low-rank E next year.
Do you mean that we should keep that one as well?
Perhaps just remove inversion with full `R` and remove the rest?

**Comment by geirev** [comment]
In Equinor, "we" will use the exact inversion (0) and the inversion with a low-rank E (3).

Using a full-rank R (1) may interest other users of the inversion code, so if this is a priority, we should keep it.
Using a low-rank R computed from E (2) is useless, as using option (3) is faster and gives precisely the same result.

Thus, we can keep options (0, 1, and 3) in the iterative smoother code but deactivate option 1 in ERT.

**Comment by dafeda** [comment]
Thank you, much appreciated.

From my understanding, have not been supporting full-rank R (1) as it is not possible to pass `R` as a parameter to the ensemble smoother.
I can add this if you deem it useful.

**Comment by geirev** [comment]
The IES routine used to be able to take a full R matrix (and it probably still is), but ert would never pass anything else than a diagonal R. I think it is good if this is still possible to do for other applications than ert.

**Comment by dafeda** [comment]
Pretty sure that the current implementation would crash if a full `R` matrix is used, but I agree that it should be possible.
I'll make a proposal that we might look at together after the holidays if you have time.

**Comment by geirev** [comment]
Sure, I am interested in a general and robust IES implementation that will serve applications other than ERT. An example is our in-house software at NORCE.

**Comment by dafeda** [comment]
I thinks that's a reasonable goal to have!

**Comment by dafeda** [comment]
I have made modifications such that it is possible to pass in a full `R` here: https://github.com/equinor/iterative_ensemble_smoother/pull/67

The API is now such that `observation_errors` can either be a 1D array representing errors (standard deviations) or a 2D array representing a covariance matrix.

The step checks the dimension of `R` and scales it by its diagonal if it is a 2D matrix.
The matrices `E`, `D` and `Y` (`response_ensemble`) are scaled by the square root of `R` as before.
Otherwise, if `R` is a 1D array, `R` is set to `None` as is never used.

```python
if len(observation_errors.shape) == 2:
    R = observation_errors
    observation_errors = np.sqrt(observation_errors.diagonal())
    R = (R.T / R.diagonal()).T
else:
    R = None
```

Does this make sense?

I've also removed the inversion `subspace_ee_r = 2` as discussed.
